### ISSUE TITLE
Dask-Arrays and Model.Store leading to kernel crash

### ISSUE THREAD

**Issue by mgrove-wwu** [hyperspy/hyperspy#2125]
Hi everyone,
as stated in the gitter chat i am experiencing a reproducable kernel crash (2 University pc's [win10] and 1 Laptop [Fedora 29]) using the hyperspy v1.4.0 and tested today the hyperspy 1.4.1 Version on my Laptop.
The issue is that using dask-arrays / lazy-signals the functions
`s.align_zero_loss_peak(subpixel=True)`
and 
`model.multifit()`
use a lot of Memory for each chung which was quite unexpectet.
Additionally the plotting is very laggy even if following the documentary (with Navigator=slider it is a Little less laggy). Now of Course each chunk has to be written into Memory which takes time but it seems to be weird still. Also the processes do not really initialize as they take a computation time of 5 minutes without any Feedback until the kernel crashes.
I wrote a script attached here to show what works and what does not.
In short though, there is no issue with everything as Long as it is not a dask-array/lazy-signal, e.g. :
`s1 = s.compute()` operating on s1 works fine while operating on s kills the python kernel which i expect to be due to Memory lack.

Thanks in Advance and i hope i described the issue well enough! Questions are welcomed.
Sincerely,
Max
[Debug_Info.zip](https://github.com/hyperspy/hyperspy/files/2804281/Debug_Info.zip)

PS: also the `model.save()` routine called leads to a kernel crashes with dask arrays due to, i guess, the memory leak.

**Comment by mgrove-wwu** [comment]
Maybe as an update and not to open a new thread: 
I happen to experience more issues with the store and save calls of a `create_model()`-object. It words for arrays generated by numpy but not for a loaded spectrum-image gathered by EELS analysis on a TITAN. For the real dataset from gatan the kernel crashes the same as stated above.
An easy example is provided with all commands - leading to the kernel crash when trying to store any model. Sadly, i think it will be my own mistake. This time it seems to not be related with dask arrays, which makes me wonder if my traceback - try was anywhere helpful...
If the metadata or additional info on the datasets i use would help, feel free to ask me here or in gitter by pm.
Sincerely,
Max


[Untitled.zip](https://github.com/hyperspy/hyperspy/files/2822075/Untitled.zip)

**Comment by mgrove-wwu** [comment]
Looking at the systemd - error messages given by `journalctl -e`hints at ctraits from cpython when the kernel dies calling `model.store()` - sorry if i am not giving the correct error dumps as i am not an expert and a self learned programmer.
```
Feb 04 10:36:04 Narnia audit[21550]: ANOM_ABEND auid=1000 uid=1000 gid=1000 ses=2 subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 pid=21550 comm="ZMQbg/1" exe="/home/max/Apps/anaconda3/envs/Hyperspy/bin/python3.7" sig=11 res=1
Feb 04 10:36:04 Narnia kernel: ZMQbg/1[21550]: segfault at 7fdf7f9cf000 ip 00007fdf7f929d46 sp 00007fffd3f6a2c0 error 4 in ctraits.cpython-37m-x86_64-linux-gnu.so[7fdf7f927000+9000]
Feb 04 10:36:04 Narnia kernel: Code: ff 15 2e a1 00 00 48 8b 55 20 48 3b 15 93 a9 00 00 48 89 43 18 0f 84 c9 02 00 00 31 c0 48 8d 0d 80 a9 00 00 83 c0 01 48 63 f8 <48> 3b 14 f9 75 f4 ff 15 fe a0 00 00 48 8b 55 28 48 3b 15 23 a8 00
Feb 04 10:36:04 Narnia systemd[1]: Started Process Core Dump (PID 21452/UID 0).
Feb 04 10:36:04 Narnia audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=systemd-coredump@2-21452-0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Feb 04 10:36:04 Narnia systemd-coredump[21453]: Removed old coredump core.ZMQbg\x2f1.1000.f4615e604ff44ff1b8f87982c312e78b.862.1549030553000000.lz4.
Feb 04 10:36:05 Narnia abrt-dump-journal-core[1344]: Abrufen der benötigten Journald-Informationen fehlgeschlagen
Feb 04 10:36:21 Narnia systemd-coredump[21453]: Process 21550 (ZMQbg/1) of user 1000 dumped core.
                                                
                                                Stack trace of thread 21550:
                                                #0  0x00007fdf7f929d46 n/a (/home/max/Apps/anaconda3/envs/Hyperspy/lib/python3.7/site-packages/traits/ctraits.cpython-37m-x86_64-linux-gnu.so)
Feb 04 10:36:21 Narnia audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=systemd-coredump@2-21452-0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'

```

**Comment by mgrove-wwu** [comment]
Now, maybe or as it seems the kernel crash only depends on my implementation of the functions using the given template in the hyperspy doc. My guess is that it has something to do with optional arguments i have to specify to be able to store a multidimensional fitted model with my components as e.g. 
```
        # Background and Convolution
        self.isbackground = False
        self.convolved = True
```
My implementation for a Lorentzian that i am currently using leads to some of the described kernel crashes:
```
#Component template from Hyperspy Documentary
#Example shows the implementation of a lorentzian fitting class
#Additionally, one can use iminuit for asymmetric errors 
#and implement other fitting routines
#e.g. using maximum likelihood-fitting

from hyperspy.component import Component

class Cauchy_Lorentz_Distribution(Component):

    def __init__(self, height=1000000., centre=0., fwhm=1.):
        """
        Comments missing
        """
        # Define the parameters
        Component.__init__(self, ('height', 'centre', 'fwhm'))

        # Optionally we can set the initial values
        self.height.value = height
        self.centre.value = centre
        self.fwhm.value = fwhm
        
        # The units (optional)
        self.height.units = 'Counts'
        self.centre.units = 'eV'
        self.fwhm.units = 'eV'

        # Once defined we can give default values to the attribute
        # For example we fix the attribure_1 (optional)
        # self.centre.free = False
        # self.height.free = False

        # And we set the boundaries (optional)
        self.height.bmin = 100.
        self.height.bmax = None
        self.centre.bmin = 0.
        self.centre.bmax = 100.
        self.fwhm.bmin = 0.
        self.fwhm.bmax = 10.
        
        # Optionally, to boost the optimization speed we can also define
        # the gradients of the function we the syntax:
        # self.parameter.grad = function
        self.height.grad = self.grad_height
        self.centre.grad = self.grad_centre
        self.fwhm.grad = self.grad_fwhm
        
    # Define the function as a function of the already defined parameters,
    # x being the independent variable value
    # Function resembles a simple gaussion
    def function(self, x):

        p1 = self.height.value
        p2 = self.centre.value
        p3 = self.fwhm.value/2

        return p1 / (((x - p2) / p3)**2 + 1)

    # Class can be extended by additional functional definitions easily
    # Just add parameters or use existing and reevaluate constraints
    
    
    # Optionally define the gradients of each parameter
    # For EELS-plasmon-fitting all parameters to be fitted are expected to be constant
    # Thus, the gradient will be set to 0
    def grad_height(self, x):
        """
        Returns d(function)/d(parameter_1) = 1 / (((x - p2) / p3)**2 + 1)
        """
        p1 = self.height.value
        p2 = self.centre.value
        p3 = self.fwhm.value/2
        return 1 / (((x - p2) / p3)**2 + 1)

    def grad_centre(self, x):
        """
        Returns d(function)/d(parameter_2) = 2 * p1 * p3**2 * (x - p2) / (((x - p2)**2 + p3**2)**2
        """
        p1 = self.height.value
        p2 = self.centre.value
        p3 = self.fwhm.value/2
        return 2 * p1 * p3**2 * (x - p2) / ((x - p2)**2 + p3**2)**2

    def grad_fwhm(self, x):
        """
        Returns d(function)/d(parameter_3) = 2 * p1 * p3 * (x - p2)**2 / (((x - p2)**2 + p3**2)**2
        """
        p1 = self.height.value
        p2 = self.centre.value
        p3 = self.fwhm.value/2
        return 2 * p1 * p3 * (x - p2)**2 / ((x - p2)**2 + p3**2)**2
```
So when using internally implemented gaussians or whatever the `model.store()` and `model.save()` routines work smoothly, but using the self defined component does not. Do you guys know why that is?

While i can use internal lorentzian and such now i would still like to implement the gamma distribution as for my fitting routine the drude model for example would not perfectly fit the the asymmetry of the plasmon shape, so i would like to store models with self implemented components which i cannot currently?

**Comment by ericpre** [comment]
Could you write your custom component as an `Expression` component (see http://hyperspy.org/hyperspy-doc/current/user_guide/model.html#specifying-custom-components)?
Even if this doesn't solve the issue here, this is the recommended way of doing it because there are a few advantages to prefer this approach.

**Comment by ericpre** [comment]
If you want to increase the chance of getting help on your issue, try to make the issue as simple as possible in order to narrow down the issue. Your may need to go a few steps back to find out where the issue is before going into something too complicated which is not manageable anymore.
At the moment, this looks fairly confusing and sadly we don't always have the required time to sort it out.

**Comment by ericpre** [comment]
I had a further look at the example in the attachments (Debug_Info.zip and Untitled.zip) you provided and there may be a few things mixed up and I can't reproduce any of the kernel crashes mentioned here.

A few points on this issue:

1. In case, this helps, there is now an expression-based Lorentzian component: https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/_components/lorentzian.py from #2134.

2. the `align_zero_loss_peak` method seems to be using too much memory for a lazy signal, see the following example to reproduce:
```python
from hyperspy._lazy_signals import LazyEELSSpectrum
s = LazyEELSSpectrum(da.random.random((300, 200, 2048), chunks=(100, 100, 2048)))
s.align_zero_loss_peak(subpixel=True, signal_range=(-10.,10.))
```

3. The kernel crash could be a red herring for a memory error.

**Comment by mgrove-wwu** [comment]
Hi epicpre,
sorry for the late answer and thank you for your effort. It seemed to be the usage of my self implemented lorentzian in addition to another problem with loading the data and not with the usage of dask. 
For now i am using already implemented models in Hyperspy which is working quite nicely with the lazy signals as well as the standard signals. In addition i tried some sympy expressions and they worked well, too with the issue that binning seems to sometimes have a bad influence on fitting due to gatan and hyperspy using different binning approaches? (It is mentioned like this at least in your user documentation)

So it seems you can close this thread. The align_zero_loss_peak method works like a charm now.

Also i cant figure out how to implement a Gamma distribution and how to adjust the threshold of the samfire-tolerance attribute well, but thats another case of my incompetence.

Last but not least samfire works for me with ipykernel too. 
So thank you anyways for your help!

Sincerely,
Max
